src/lib/matcher.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * This module provides the basic type machinery to build pipeline
+ * translators efficiently and safely.
+ *
+ * It exposes 2 main building blocks:
+ * - `StepMatcher`: a type mapping each available pipeline steps to a
+ *   translation function
+ * - `matchStep`: a function accepting a `StepMatcher` and that return a function
+ *   transforming a pipeline step to an arbiratry output.
+ *
+ * This system is supposed to be typesafe, which means that if you add a
+ * new type of `PipelineStep`, the typescript compiler is expected to fail
+ * if your matcher implementation doesn't provide a mapping for this new kind
+ * of step.
+ */
+import { PipelineStep, PipelineStepName } from './steps';
+
+/**
+ * `StepByType` is a conditional type that expects the `PipelineStep` type and
+ * a valid step name and return the only possible step type matching this name.
+ *
+ * For instance, `StepByType<PipelineStep, 'delete'>` is equivalent to `DeleteStep`
+ */
+export type StepByType<A, T> = A extends { name: T } ? A : never;
+
+/**
+ * `StepMatcher` is a type built dynamically according to each available step
+ * name that associates each of this step name to a specific processing function
+ * accepting a step of corresponding type. For instance, it will generate the
+ * `rename: (step: RenameStep) => T` mapping entry automatically.
+ *
+ * This type will help to enforce that each kind of step has a corresponding
+ * translation function at compile time.
+ */
+export type StepMatcher<T> = { [K in PipelineStepName]: (step: StepByType<PipelineStep, K>) => T };
+
+/**
+ * Helper exception to handle unreachable conditions in a switch statement.
+ * This will help the typescript compiler to raise an error if one or several
+ * kind of steps are missing in a switch / case statement.
+ *
+ */
+class Unreachable extends Error {
+  constructor(value: never) {
+    super(`value ${value} is not valid`);
+  }
```

Comment:
I meant `unreachable` as in [unreachable code](https://en.wikipedia.org/wiki/Unreachable_code).

It really is a way to give a hint to the compiler. While "compiling" the different `swtich / case` branches, typescript will narrow down the possible types (string values here are actually types). If some `cases` are not written down (e.g. `my-step-name`), when typescript will analyze the "unreachable" branch, it will complain about an `my-step-name` being incompatible with the `never` type.

Reply:
Ah OK! I haven't thought of that type of unreachability.
It still feels weird to have code that will appear on the final bundle to throw errors that we only need at compile time.

Why shouldn't we just define an interface then:
```
interface Matcher {
    domain(): ... 
    filter(): ...
   ...
}
```
That seems more idiomatic and will provide simple messages saying '<method> is not implemented'
Also that means transforming matchers into classes that extends the Matcher interface.

Reply:
Cool then! :100: Thanks for the explanations